server/src/main/java/org/jboss/as/server/deployment/module/descriptor/DeploymentStructureDescriptorParser.java

```diff
@@ -238,9 +240,25 @@ public void deploy(final DeploymentPhaseContext phaseContext) throws DeploymentU
         }
     }
 
-    private void handleDeployment(final DeploymentPhaseContext phaseContext, final DeploymentUnit deploymentUnit, final ModuleSpecification moduleSpec, final ModuleStructureSpec rootDeploymentSpecification, Set<ModuleIdentifier> additionalModules) throws DeploymentUnitProcessingException {
+    private void handleDeployment(final DeploymentPhaseContext phaseContext, final DeploymentUnit deploymentUnit, final ModuleSpecification moduleSpec, final ModuleStructureSpec rootDeploymentSpecification, final Set<AdditionalModuleSpecification> additionalModules) throws DeploymentUnitProcessingException {
         final Map<VirtualFile, ResourceRoot> resourceRoots = resourceRoots(deploymentUnit);
-        moduleSpec.addUserDependencies(rootDeploymentSpecification.getModuleDependencies());
+        // handle unmodifiable module dependencies with alias
+        List<ModuleDependency> moduleDependencies = rootDeploymentSpecification.getModuleDependencies();
+        List<ModuleDependency> copy = new ArrayList<>();
+        List<ModuleDependency> aliasDependencies = new ArrayList<>();
+        copy.addAll(moduleDependencies);
+        Iterator<ModuleDependency> it = copy.iterator();
+        while (it.hasNext()) {
+            ModuleDependency dependency = it.next();
+            ModuleIdentifier identifier = dependency.getIdentifier();
+            for (AdditionalModuleSpecification module : additionalModules) {
+                if (module.getAliases().contains(identifier)) {
+                    aliasDependencies.add(new ModuleDependency(dependency.getModuleLoader(), module.getModuleIdentifier(), dependency.isOptional(), dependency.isExport(), dependency.isImportServices(), dependency.isUserSpecified()));
+                }
+            }
+        }
+        copy.addAll(aliasDependencies);
+        moduleSpec.addUserDependencies(copy);
```

Comment:
This all seems a bit overly complicated; what is this code trying to accomplish?

Reply:
Iterator is not necessary, I changed to for-each.
This only adds real module dependency to moduleSpec if matched alias is found.